src/cmdstan/stansummary_helper.hpp

```diff
@@ -463,129 +416,56 @@ void write_header(const std::vector<std::string> &header,
  * Output statistics for a set of parameters
  * either as fixed-width text columns or in csv format.
  *
- * @param in set of samples from one or more chains
- * @param in matrix of statistics
- * @param in vector of output column widths
- * @param in vector of output column formats
- * @param in size of longest parameter name - (width of 1st output column)
- * @param in significant digits required
- * @param in vector of column indexes to output from chains
- * @param in output format flag:  true for csv; false for plain text
- * @param in output stream
+ * @param param_names vector of requested parameter names
+ * @param stats matrix of computed statistics
+ * @param col_widths vector of output column widths
+ * @param col_formats vector of output column formats
+ * @param max_name_length longest parameter name - (width of 1st output column)
+ * @param sig_figs significant digits required
+ * @param as_csv flag - true for csv; false for plain text
+ * @param reorder_params flag - if true, reorder param names
+ * @param out output stream
  */
-void write_params(const stan::mcmc::chains<> &chains,
-                  const Eigen::MatrixXd &params,
-                  const Eigen::VectorXi &col_widths,
-                  const Eigen::Matrix<std::ios_base::fmtflags, Eigen::Dynamic,
-                                      1> &col_formats,
-                  int max_name_length, int sig_figs, std::vector<int> cols,
-                  bool as_csv, std::ostream *out) {
-  int i = 0;
-  for (int i_chains : cols) {
+void write_stats(const std::vector<std::string> &param_names,
+                 const Eigen::MatrixXd &stats,
+                 const Eigen::VectorXi &col_widths,
+                 const Eigen::Matrix<std::ios_base::fmtflags, Eigen::Dynamic, 1>
+                     &col_formats,
+                 int max_name_length, int sig_figs, bool as_csv,
+                 bool reorder_params, std::ostream *out) {
+  auto pnames = param_names;
+  if (reorder_params) {
+    pnames = order_param_names_row_major(param_names);
+  }
```

Comment:
If I am reading this correctly, it is only reordering the names, not the actual numerical values?

Before:

```
Inference for Stan model: issue_342_model
1 chains: each with iter=(10); warmup=(0); thin=(1); 10 iterations saved.

Warmup took 0.23 seconds
Sampling took 0.0018 seconds

                Mean   MCSE  StdDev       5%   50%   95%  N_Eff   N_Eff/s  R_hat

lp__             -15    1.2     3.7 -2.6e+01   -15   -13     10   5.7e+03   0.92
accept_stat__   0.69   -nan    0.30     0.20  0.86   1.0    -45  -2.6e+04   0.92
stepsize__      0.89    nan    0.00     0.89  0.89  0.89    nan       nan    nan
treedepth__     0.90   0.10    0.32     0.00   1.0   1.0     10   5.7e+03    nan

x[1]            0.61   -nan    0.22  2.8e-01  0.67  0.98    -48  -2.7e+04   0.89
x[2]            0.48  0.051    0.16  1.8e-01  0.48  0.77     10   5.7e+03   0.99
y[1,1]          0.43   0.12    0.39  2.3e-02  0.17  0.93     10   5.7e+03    1.1
y[1,2]          0.60  0.076    0.24  2.5e-01  0.71  0.88     10   5.7e+03   0.90
y[1,3]          0.47   -nan    0.33  4.5e-02  0.71  0.94    -92  -5.2e+04   0.90
y[2,1]          0.51   -nan    0.27  1.7e-01  0.49  0.93   -354  -2.0e+05   0.91
y[2,2]          0.56  0.087    0.28  1.6e-01  0.65  0.99     10   5.7e+03   0.94
y[2,3]          0.52  0.100    0.32  1.9e-03  0.65  0.93     10   5.7e+03    1.2

Samples were drawn using hmc with nuts.
For each parameter, N_Eff is a crude measure of effective sample size,
and R_hat is the potential scale reduction factor on split chains (at 
convergence, R_hat=1).
```

This PR:
```
Inference for Stan model: issue_342_model
1 chains: each with iter=10; warmup=1000; thin=1; 10 iterations saved.

Warmup took 0.23 seconds
Sampling took 0.0018 seconds

                Mean   MCSE   StdDev   MAD     5%   50%   95%  ESS_bulk  ESS_tail  R_hat

lp__             -15    1.2  3.7e+00  0.78    -21   -15   -13       5.0       5.0    1.3
accept_stat__   0.69  0.096  3.0e-01  0.25   0.21  0.78  0.99
stepsize__      0.89    nan  1.2e-16  0.00   0.89  0.89  0.89
treedepth__     0.90   0.10  3.2e-01  0.00   0.45   1.0   1.0

x[1]            0.61  0.069  2.2e-01  0.11   0.28  0.67  0.87       5.0       5.0   0.91
x[2]            0.48  0.051  1.6e-01  0.16   0.26  0.45  0.69       5.0       5.0    1.0
y[1,1]          0.43   0.12  3.9e-01  0.15  0.059  0.16  0.93       5.0       5.0    1.4
y[1,2]          0.51  0.087  2.7e-01  0.38   0.17  0.49  0.90       5.0       5.0    1.4
y[1,3]          0.60  0.076  2.4e-01  0.18   0.25  0.66  0.85       5.0       5.0    1.1
y[2,1]          0.56  0.087  2.8e-01  0.31   0.23  0.53  0.93       5.0       5.0   0.92
y[2,2]          0.47   0.10  3.3e-01  0.39  0.068  0.48  0.86       5.0       5.0   0.99
y[2,3]          0.52  0.100  3.2e-01  0.21  0.036  0.54  0.93       5.0       5.0    1.3

Samples were drawn using hmc with nuts.
For each parameter, ESS_bulk and ESS_tail measure the effective sample size 
for the entire sample (bulk) and for the the .05 and .95 tails (tail), 
and R_hat measures the potential scale reduction on split chains.
At convergence R_hat will be very close to 1.00.
```

The variable names are in the same column, the means etc are jumbled

This also means we apparently have no tests that check that the correct name is actually being paired with the correct mean, etc

Reply:
the names are what is used to generate the stats.  
https://github.com/stan-dev/cmdstan/blob/04b00166ba78616840925e4040b7ebabd73d1b71/src/cmdstan/stansummary_helper.hpp#L365-L370

under the hood, the chainset object translates names to indices.
https://github.com/stan-dev/stan/blob/9048555070493679287a25086e7b85b72ad9531f/src/stan/mcmc/chainset.hpp#L133-L148

in the above output, where are things jumbled?

Reply:
this time for sure!

```
Inference for Stan model: issue_342_model
1 chains: each with iter=10; warmup=1000; thin=1; 10 iterations saved.

Warmup took 0.23 seconds
Sampling took 0.0018 seconds

                Mean   MCSE  StdDev   MAD     5%   50%   95%  ESS_bulk  ESS_tail  R_hat

lp__             -15    1.2     3.7  0.78    -21   -15   -13       5.0       5.0    1.3
accept_stat__   0.69  0.096    0.30  0.25   0.21  0.78  0.99
stepsize__      0.89    nan    0.00  0.00   0.89  0.89  0.89
treedepth__     0.90   0.10    0.32  0.00   0.45   1.0   1.0

x[1]            0.61  0.069    0.22  0.11   0.28  0.67  0.87       5.0       5.0   0.91
x[2]            0.48  0.051    0.16  0.16   0.26  0.45  0.69       5.0       5.0    1.0
y[1,1]          0.43   0.12    0.39  0.15  0.059  0.16  0.93       5.0       5.0    1.4
y[1,2]          0.60  0.076    0.24  0.18   0.25  0.66  0.85       5.0       5.0    1.1
y[1,3]          0.47   0.10    0.33  0.39  0.068  0.48  0.86       5.0       5.0   0.99
y[2,1]          0.51  0.087    0.27  0.38   0.17  0.49  0.90       5.0       5.0    1.4
y[2,2]          0.56  0.087    0.28  0.31   0.23  0.53  0.93       5.0       5.0   0.92
y[2,3]          0.52  0.100    0.32  0.21  0.036  0.54  0.93       5.0       5.0    1.3
```

Reply:
the reordering needs to happen before the stats are computed; not when the stats are written.  fixed code and added unit test for this.